File: 02/02a.py
import sys

if __name__ == "__main__":
    MAX_COUNTS = {
        "red": 12,
        "green": 13,
        "blue": 14,
    }
    answer = 0
    for game in sys.stdin:
        valid = True
        game_decriptor, outcomes = game.rstrip().split(":")
        _, game_id = game_decriptor.split()
        for outcome in outcomes.split("; "):
            for colorpair in outcome.split(", "):
                count, color = colorpair.split()
                if int(count) > MAX_COUNTS[color]:
                    valid = False
                    break
        if valid:
            answer += int(game_id)
    print(answer)

File: haskell/02/02a.hs
import Data.List.Split

main :: IO ()
main = do
  contents <- getContents
  print $ sum $ map validateGames (lines contents)

validateGames line =
  let [game_name, games] = splitOn ": " line
      [_, game_id] = words game_name
   in if all validate (splitOn ";" games)
        then read game_id
        else 0

validate game =
  let colors = splitOn ", " game
   in all validateColor colors

validateColor line =
  let [countL, color] = words line
      count = read countL :: Int
   in case color of
        "red" -> count <= max_red
        "green" -> count <= max_green
        "blue" -> count <= max_blue
  where
    max_red = 12
    max_green = 13
    max_blue = 14